velox/type/DecimalUtil.h

```diff
@@ -32,5 +32,39 @@ class DecimalUtil {
   /// Helper function to convert a decimal value to string.
   template <typename T>
   static std::string toString(const T& value, const TypePtr& type);
+
+  inline static int128_t rescaleWithRoundUp(
+      const int128_t& unscaledValue,
+      const uint8_t fromScale,
+      const std::pair<uint8_t, uint8_t>& toPrecisionScale,
+      bool& isNullOut) {
+    auto rescaledValue = unscaledValue;
+    auto rf = toPrecisionScale.second - fromScale;
+    if (rf >= 0) {
+      rescaledValue = unscaledValue * DecimalUtil::kPowersOfTen[rf];
+    } else {
+      rf = -1 * rf;
+      rescaledValue = unscaledValue / DecimalUtil::kPowersOfTen[rf];
+      int128_t remainder = unscaledValue % DecimalUtil::kPowersOfTen[rf];
+      if (unscaledValue >= 0) {
+        if (remainder > DecimalUtil::kPowersOfTen[rf] / 2) {
+          rescaledValue++;
+        }
+      } else {
+        if (remainder < -DecimalUtil::kPowersOfTen[rf] / 2) {
+          rescaledValue--;
+        }
+      }
+    }
+    // Check overflowing
+    auto valueToCompare = rescaledValue;
+    if (rescaledValue < 0) {
```

Comment:
Wrap this into VELOX_DCHECK also?
`VELOX_DCHECK_LE(valueToCompare, DecimalUtil::kPowersOfTen[toPrecisionScale.first]);`
`VELOX_DCHECK_GE(valueToCompare, -DecimalUtil::kPowersOfTen[toPrecisionScale.first]);`

Reply:
Since we do need the if check, you can add another condition for negation
```
valueToCompare > DecimalUtil::kPowersOfTen[toPrecisionScale.first] || 
valueToCompare < -DecimalUtil::kPowersOfTen[toPrecisionScale.first]
```